Content.Server/Revenant/EntitySystems/RevenantSystem.Abilities.cs

```diff
@@ -0,0 +1,302 @@
+            return;
+
+        args.Handled = true;
+
+        var lookup = _lookup.GetEntitiesInRange(uid, component.DefileRadius, LookupFlags.Approximate | LookupFlags.Anchored);
+
+        for (var i = 0; i < component.DefileTilePryAmount; i++)
+        {
+            //get random coordinates in the radius (technically a square but shut up)
+            var coords = new EntityCoordinates(uid,
+                (_random.NextFloat(-component.DefileRadius, component.DefileRadius), _random.NextFloat(-component.DefileRadius, component.DefileRadius)));
+
+            var gridID = coords.GetGridUid(EntityManager);
+            if (_mapManager.TryGetGrid(gridID, out var map))
+                map.GetTileRef(coords).PryTile(_mapManager, entityManager: EntityManager, robustRandom: _random);
+        }
+
+        foreach (var ent in lookup)
+        {
+            //break windows
+            if (HasComp<TagComponent>(ent) && _tag.HasAnyTag(ent, "Window"))
+            {
+                //hardcoded damage specifiers til i die.
+                var dspec = new DamageSpecifier();
+                dspec.DamageDict.Add("Structural", 15);
+                _damage.TryChangeDamage(ent, dspec);
+            }
+
+            //randomly opens some lockers and such.
+            if (_random.Prob(component.DefileEffectChance) && HasComp<EntityStorageComponent>(ent))
+                _entityStorage.OpenStorage(ent);
+
+            //chucks shit
+            if (_random.Prob(component.DefileEffectChance) && HasComp<ItemComponent>(ent) &&
+                TryComp<PhysicsComponent>(ent, out var phys) && phys.BodyType != BodyType.Static)
+                _throwing.TryThrow(ent, _random.NextAngle().ToWorldVec());
+
+            //flicker lights
+            if (_random.Prob(component.DefileEffectChance) && HasComp<PoweredLightComponent>(ent))
+                RaiseLocalEvent(ent, new GhostBooEvent());
+        }
+    }
+
+    private void OnOverloadLightsAction(EntityUid uid, RevenantComponent component, RevenantOverloadLightsActionEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (!CanUseAbility(uid, component, component.OverloadCost, component.OverloadDebuffs))
+            return;
+
+        args.Handled = true;
+
+        var poweredLights = GetEntityQuery<PoweredLightComponent>();
+        var lookup = _lookup.GetEntitiesInRange(uid, 5).ToArray();
+
+        for (var i = 0; i < lookup.Length; i++)
+        {
+            var ent = lookup[i];
+
+            if (!poweredLights.TryGetComponent(ent, out var light))
+                continue;
+
+            var ev = new GhostBooEvent(); //light go flicker
+            RaiseLocalEvent(ent, ev);
+
+            if (_random.Prob(component.OverloadBreakChance))
+            {
+                //values
+                _explosion.QueueExplosion(ent, "RevenantElectric", 15, 3, 5, canCreateVacuum: false);
+            }
+        }
```

Comment:
How come this is a for loop but the other one was foreach? You can just queue explosions until the end which also means you can avoid the ToArray.

Reply:
iirc the reason this was a for-loop instead of foreach is that sometimes one of the entities would be destroyed from the explosion, which would cause an error for the collection being modified.